metagraph/src/graph/annotated_graph_algorithm.cpp

```diff
@@ -15,249 +12,417 @@ namespace graph {
+
+            size_t size = end - begin;
+            size_t label_in_cutoff = std::ceil(config.label_mask_in_unitig_fraction * size);
+            if (in_kmer_count < label_in_cutoff)
+                return {};
+
+            size_t out_kmer_count = 0;
+            size_t other_kmer_count = 0;
+            size_t label_out_cutoff = std::floor(config.label_mask_out_unitig_fraction * size);
+            size_t other_cutoff = std::floor(config.label_mask_other_unitig_fraction * size);
+
+            for (size_t i = begin; i < end; ++i) {
+                if (counts[2 * path[i] + 1] > max_label_out_count
+                        && ++out_kmer_count > label_out_cutoff) {
+                    return {};
+                }
+
+                if (check_other && other_mask[path[i]]
+                        && ++other_kmer_count > other_cutoff) {
+                    return {};
+                }
+            }
+
+            return { std::make_pair(begin, end) };
+        },
+        num_threads
+    );
+
+    return masked_graph;
+}
+
+
+/**
+ * Helpers
+ */
+
+std::shared_ptr<MaskedDeBruijnGraph>
+make_initial_masked_graph(std::shared_ptr<const DeBruijnGraph> graph_ptr,
+                          sdsl::int_vector<> &counts,
+                          sdsl::bit_vector&& mask,
+                          bool add_complement,
+                          size_t num_threads) {
+    // counts is a double-length vector storing the in-label and out-label
+    // counts interleaved
+    assert(counts.size() == mask.size() * 2);
+
+    add_complement |= graph_ptr->get_mode() == DeBruijnGraph::CANONICAL;
+    auto masked_graph = std::make_shared<MaskedDeBruijnGraph>(
+        graph_ptr,
+        add_complement
+            ? std::make_unique<bitmap_vector>(mask)
+            : std::make_unique<bitmap_vector>(std::move(mask)),
+        true
+    );
+
+    logger->trace("Constructed masked graph with {} nodes", masked_graph->num_nodes());
+
+    if (add_complement) {
+        logger->trace("Adding reverse complements");
+        std::mutex vector_backup_mutex;
+        std::atomic_thread_fence(std::memory_order_release);
+        masked_graph->call_sequences([&](const std::string &seq, const std::vector<node_index> &path) {
+            std::string rc_seq = seq;
+            std::vector<DeBruijnGraph::node_index> rc_path = path;
+            reverse_complement_seq_path(*graph_ptr, rc_seq, rc_path);
+
+            auto it = rc_path.rbegin();
+            for (size_t i = 0; i < path.size(); ++i, ++it) {
+                if (*it) {
+                    uint64_t in_count = atomic_fetch(counts, path[i] * 2, vector_backup_mutex, MEM_ORDER);
+                    uint64_t out_count = atomic_fetch(counts, path[i] * 2 + 1, vector_backup_mutex, MEM_ORDER);
+                    atomic_fetch_and_add(counts, *it * 2, in_count, vector_backup_mutex, MEM_ORDER);
+                    atomic_fetch_and_add(counts, *it * 2 + 1, out_count, vector_backup_mutex, MEM_ORDER);
```

Comment:
Why do you need to pass a mutex? Isn't it always allowed to do atomic_fetch and add with any memory order?

Reply:
Atomic operations on unaligned blocks of size > 64 are not allowed. So if the int_vector width is too large, this defaults to doing all operations with a lock.

Reply:
how can it be larger than 64 for int_vector?

Reply:
All of the operations need to at least be aligned to byte blocks. So, if for example, your width is 63, then you need to have atomic access to an unaligned block of >64-bits since the starting point has to be byte-aligned.

Reply:
got it